### PR TITLE
feat: restructure feature-flags as object instead of being an array...

### PR DESCRIPTION
**Pull request by rbattala-ck**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -79,26 +79,25 @@ The schema for the `toggles` config:
...
         "toggles": {
-            "type": "array",
-            "items": {
-                "type": "object",
-                "properties": {
-                    "id": {
-                        "type": "string"
+            "type": "object",
+            "patternProperties": {
+                "^[A-Za-z0-9 -_.]": {
```
Comment (nick-penaranda-ck) [inline README.md:84]: This only matches the first character, do we want `*$` at the end?

File: src/main/schema.ts [hunk 2]
```diff
@@ -1,29 +1,19 @@
...
-                    fraction: {
-                        type: 'number',
-                        minimum: 0.0,
-                        maximum: 1.0,
-                    },
-                    comment: {
-                        type: 'string',
-                    },
+    patternProperties: {
+        '^[A-Za-z0-9 -_.]': {
```
Comment (nick-penaranda-ck) [inline src/main/schema.ts:4]: Commented on the readme but really it applies here: This is only testing the first character; if we want to constrain the entire property name to this character set, need `*$` at the end

File: src/main/toggles.ts [hunk 3]
```diff
@@ -13,36 +14,37 @@ import { memoize, objectMatchesSchema } from './utils'
...
+                    }
                 })
+                resolve(Promise.resolve(rawToggles))
 
                 resolve(Promise.resolve(rawToggles))
 
             } else {
-                defaultLogger(['error', 'toggleMap'], `Value of 'toggles' should be an array`)
-                reject(new Error(`Value of 'toggles' should be an array`))
+                defaultLogger(['error', 'toggleMap'], `Value of 'toggles' should be an object`)
```
Comment (nick-penaranda-ck) [inline src/main/toggles.ts:31]: DRY: error message constant

File: README.md [hunk 4]
```diff
@@ -79,26 +79,25 @@ The schema for the `toggles` config:
...
+                    "type": "object",
+                    "properties": {
+                        "fraction": {
+                            "type": "number",
+                            "minimum": 0.0,
+                            "maximum": 1.0,
+                        },
+                        "type": {
+                            "type": "string",
```
Comment (edwin-kim-ck) [inline README.md:93]: I think it would be helpful to change this to an enum of ["RAMP", "BOOL"].

File: src/main/schema.ts [hunk 5]
```diff
@@ -1,29 +1,19 @@
...
+            type: 'object',
+            properties: {
+                fraction: {
+                    type: 'number',
+                    minimum: 0.0,
+                    maximum: 1.0,
+                },
+                type: {
+                    type: 'string',
```
Comment (edwin-kim-ck) [inline src/main/schema.ts:13]: Would there be any thought to making this an enum like ["BOOL", "RAMP"]? Or would the client need to have to be able to define this?

File: src/main/toggles.ts [hunk 6]
```diff
@@ -13,36 +14,37 @@ import { memoize, objectMatchesSchema } from './utils'
...
-                toggles.forEach((next: IToggleDescription) => {
-                    rawToggles.set(next.id, next)
+        config().watch<IToggleDescription>(togglePath).onValue(({toggles}): void => {
+            if (objectMatchesSchema(toggleSchema, toggles)) {
+                Object.keys(toggles).forEach((key) => {
+                    if (toggles[key]) {
+                        rawToggles.set(key, toggles[key])
+                    }
                 })
+                resolve(Promise.resolve(rawToggles))
```
Comment (edwin-kim-ck) [inline src/main/toggles.ts:26]: There seems to be an extra resolve here.

File: src/main/types.ts [hunk 7]
```diff
@@ -1,10 +1,18 @@
...
-    id: string
+export interface IToggleDetails {
+    type: string
     fraction: number
     description?: string
     comment?: string
 }
 
+export interface IToggle {
+    [key: string]: IToggleDetails
```
Comment (edwin-kim-ck) [inline src/main/types.ts:11]: Curious about this. I am guessing that this was changed so a toggle could be looked up by key. Is using the current structure, an array of toggles possible? If the type were added to the original IToggleDescription and made optional would it be possible to not be a breaking change?

File: src/main/schema.ts (hunk 5 above)
Comment (rbattala-ck) [reply src/main/schema.ts:13]: Not sure how this can be achieved. But will be good to have that check.

File: src/main/toggles.ts (hunk 6 above)
Comment (rbattala-ck) [reply src/main/toggles.ts:26]: 👍

**Pull request merged**